Fill.py

```diff
@@ -452,17 +454,106 @@ def distribute_early_items(multiworld: MultiWorld,
 
         fill_locations.extend(early_locations)
         multiworld.random.shuffle(fill_locations)
+
+    fill_locations = [location for location in original_order if not location.item]
+
     return fill_locations, itempool
 
 
+def balanced_shuffle(multiworld: MultiWorld, fill_locations: list[Location], itempool: list[Item]) -> list[Location]:
+    balancing_factor = 0.5  # This would be a setting instead. Acts as a percentage. 0.0 is min, 1.0 is max.
+
+    # First, we shuffle the location pool.
+    multiworld.random.shuffle(fill_locations)
+
+    # If balancing factor is 0, don't do any more unnecessary work.
+    if balancing_factor == 0.0:
+        return fill_locations
+
+    # Balancing only has an effect if there is progression
+    amount_of_progression = sum(1 for item in itempool if item.advancement)
+    if not amount_of_progression:
+        return fill_locations
+
+    # If balancing factor is not 0, we split up the locations list by players.
+    locations_per_player = collections.defaultdict(list)
+    for location in fill_locations:
+        locations_per_player[location.player].append(location)
+
+    # shuffle the player order in the dict so that no player gets an inherent advantage.
+    players_with_locations = list(locations_per_player)
```

Comment:
```suggestion
    players_with_locations = list(locations_per_player.keys())
```
This makes it easier to understand for a human reading the code.
(You did the same thing below.)